sherpa/plot/bokeh_backend.py

```diff
@@ -0,0 +1,829 @@
+        nrows : int
+            The number of rows.
+        ncols : int
+            The number of columns.
+        create : bool, optional
+            If True then create the plots
+        top : int
+            The row that is set to the ratio height, numbered from 0.
+        ratio : float
+            The ratio of the height of row number top to the other
+            rows.
+        """
+        # This function is written to work with an incomplete grid of plots,
+        # so there is a whole lot of "if plotnum is not None" in here.
+        # If we were to require that all grid are complete (i.e. a 3x2 grid)
+        # always has 6 plots in a fixed order, the implementation would be
+        # simpler.
+        if create:
+            figs = [self._figure() for n in range(nrows * ncols)]
+            self.current_fig = gridplot(figs, ncols=ncols)
+            self.current_axis = figs[row * ncols + col]
+
+        # space is dynamically allocated and lots of things can change the
+        # height, e.g. a user may have changed bokeh defaults.
+        # It's easy to change the height of a plot, but we first need to find
+        # what height we consider "normal".
+        # If we simply were to multiply the height of the top row by the ratio,
+        # then the total height of the plot get larger every time this method
+        # is called, which, for ncols > 1, might be several times.
+        # We pick the first plot in the grid that's not in row "top" and use
+        # that as the "normal" height. That might fail for more complex grids,
+        # but I declare that outside the scope of this function.
+        height = None
+        for fig in enumerate(self.current_fig.children):
+            if fig[1] != top:
+                height = fig[0].height
+                break
+        # normal_hight is None if we only have one row, so we don't need to
+        # change anything.
+        if height is not None:
+            for i in range(ncols):
+                plotnum = self._index_axis(top, i)
+                # We might have an incomplete grid of plots, and then
+                # plotnum might be None.
+                if plotnum is not None:
+                    self.current_fig[plotnum].height = height * ratio
+
+        # Axis sharing
+        for c in range(ncols):
+            basenum = None
+            for r in range(nrows):
+                index = self._index_axis(0, c)
+                if basenum is None:
+                    basenum = index
+                else:
+                    self.current_fig[index].x_range = \
+                        self.current_fig[basenum].x_range
+
+        plotnum = self._index_axis(row, col)
+        if plotnum is None:
+            raise ArgumentErr("No plot at row {} and column {}." +
+                              "Use create=True to create an entirely new, " +
+                              "complete grid of plots.".format(row, col)) from None
```

Comment:
The use of "blah {} {}" + ... + "text".format(row,col) doesn't actually pass the row/col arguments to the first bit. So this is one where a f-string change is definitely appropriate.

```suggestion
            raise ArgumentErr(f"No plot at row {row} and column {col}."
                              "Use create=True to create an entirely new, "
                              "complete grid of plots.") from None
```